File: Python/11_yet_another_sum.py
import sys


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def calculate_sum(array):
    s1 = sum(array)
    reciprocals = [1 / elem for elem in array]
    positives = sorted(filter(lambda x: x > 0, reciprocals), key=abs)
    negatives = sorted(filter(lambda x: x < 0, reciprocals), key=abs)
    return (sum(positives) + sum(negatives)) + s1


def main():
    reader = sys.stdin
    _ = next(reader)
    array = convert_to_intlist(next(reader))
    result = calculate_sum(array)
    print(result)


if __name__ == '__main__':
    main()

File: Haskell/11-yet-another-sum.hs
import Data.List (sortOn)


convertToIntList :: String -> [Int]
convertToIntList = map read . words


calculateSum :: [Int] -> Double
calculateSum array =
    let reciprocals = map (\x -> 1 / fromIntegral x) array
        positives = sortOn abs $ filter (> 0) reciprocals
        negatives = sortOn abs $ filter (< 0) reciprocals
    in (sum positives + sum negatives) + fromIntegral (sum array)


main :: IO()
main = do
    _ <- readLn :: IO Int
    array <- fmap convertToIntList getLine
    let result = calculateSum array
    print result
